### PR TITLE
feat(ark): 🎸 basic store with hooks

### PR DESCRIPTION
**Pull request by agroupp**

### CODE REVIEW

File: packages/ark/src/lib/ark/utils/interceptors-from-features.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { ArkFeatures, ArkInterceptors } from '../entities';
+
+export function interceptorsFromFeatures<State>(features: ArkFeatures<State>[]): ArkInterceptors<State> {
+  let result: ArkInterceptors<State> = { beforeInit: [], afterInit: [], beforeUpdate: [], afterUpdate: [] };
+  const interceptors = features.map(f => f.interceptors);
+  for (const interceptor of interceptors) {
+    result = { ...result, ...interceptor };
```
Comment (shyagamzo) [inline packages/ark/src/lib/ark/utils/interceptors-from-features.ts:7]: Is this the intended behavior?
Seems like each feature will override the interceptors collected from the previous ones.

Shouldn't interceptors be aggregated?

Comment (agroupp) [reply packages/ark/src/lib/ark/utils/interceptors-from-features.ts:7]: Why? It will add interceptors. I have an example in dev-app and it seems to work. Am I missing something?

**Review by OmerGronich — COMMENTED** [review]
love it ✨

File: packages/ark/src/lib/ark/utils/interceptors-from-features.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { ArkFeatures, ArkInterceptors } from '../entities';
+
+export function interceptorsFromFeatures<State>(features: ArkFeatures<State>[]): ArkInterceptors<State> {
```
Comment (shyagamzo) [inline packages/ark/src/lib/ark/utils/interceptors-from-features.ts:4]: How about [something like this](https://www.typescriptlang.org/play?ts=4.9.5#code/KYDwDg9gTgLgBAMwK4DsDGMCWEV0ymYKNYMGaAZwDEoIBbK4AQxiSmAoB4BlGF4OKEIoAJhTi9+AeQBGAK2AYAfAAoEzVuwoAuOAEEoAa0Ys2HHn0JKA2gF0AlLoOGAkgSIkylC-yVwA3gBQcHDsmrjqploAdOwiSCQqKkwANikANAF47sSk5FDiAL72cAC8fkEhISqyChjRhsAAnhQq+IS5XgUlTOIqjU0QCHAwTWDAQ9kdnvkU9nb20QjQAKJMaAAWKhsQEIZlFcFVIakpZQFHxyHRN6fpl1fWO3u2utY30adPu4a2mR9tHIzSjfF5wAD84LgC1sDxChQA3A9ikiHmE2LhTkj4Zl-HAZMBluw3JgYG8-nAmAgOiSydCKQSicAAKpgET8cmZKkdVnswjkuCFSniZxuaZ5bySKz2JGFQKBdpEBDrASinLKkg+QiCEDCMQSSzAWqKGCHEJgWhgXRS4CBOWBUbjA3SeQm84AJUU0BEnAGkyYKCamVQhhQEAA7iglKjQJBYCMxgIAEKE6DAWlijwSqBagRCYCicQ243Kc6A0mYVI262GkrlZ2EJGx6DwR2q6lEDNA7O5nV6ouGkumssUQ01-h1vwANwgmBETfALYTTpTTN5-EznXyvfzhYbRtdpdKcBURzQbHYBGr+-uIRQwHD14ACkxYJWUrmlLe4KO004jJuGrAJ+9yTnAL5vqkn4LnGraJvoHZQOuhCbsCOY2n2Bb6sWh7DseKijhyN4-vkwD-q46oqp+YEznOMaLvGipQEB+gAd2XRcBhu7YYOuFmviqbEigpK6KuaZduKXSfnY2KUohtJOPJwkwKhPY2jYtiyYyabIWRcBiewumqVJ6kyUc3JELpik8myG7sdupmaXa8rNvGBZIHQrHGBoZgANL4CIFwhMZswmJowD3ParnwExLHOGFZicAl7D+aImF7vFPkpQFmRcbqWEDi6dTDpUcCGAFujJcAqXzkcTFoTo4GvlgUFqpJsyftGzmBNFy4CCFlBVTu+V7jhxV+MemVRMBU3hTV0QDQUVW5YaXW9W2XlVZxhrpTxRUmhNcCLdQWXAepCJwAA9JdcAACpSAAIlIcBoDgWAoEgAgUMAACOn3oAI4akhscAAD6BEAA)?

File: packages/ark/src/lib/ark/ark.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+export abstract class Ark<State> implements ArkInterface<State> {
+  private readonly $state: WritableSignal<State>;
+  private readonly _hooks: Hooks<State>;
+
+  constructor(private initialState: State, ...features: ArkFeatures<State>[]) {
+    this._hooks = new Hooks(interceptorsFromFeatures<State>(features));
+
+    // Before init interceptors,
+    // here we can change the initial state
+    for (const interceptor of this._hooks.beforeInit) {
```
[the remark this answers is not in the document]
Comment (agroupp) [reply packages/ark/src/lib/ark/ark.ts:16]: I thought about it at the beginning and even started to implement. The thing is that in different parts of Ark class interceptors behave differently. So I decided to not to overcomplicate it.

File: packages/ark/src/lib/ark/entities/hooks.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+export const ARK_HOOK_GROUPS = ['beforeInit', 'afterInit'] as const;
+export type ArkHookGroup = typeof ARK_HOOK_GROUPS[number];
+
+export interface OnBeforeUpdate<State> {
+  onBeforeUpdate(currentState: State, newState: Partial<State>): Partial<State>;
```
Comment (ronnetzer) [inline packages/ark/src/lib/ark/entities/hooks.ts:5]: isn't it the same as beforeUpdate interceptor? why do we need both?

**Review by ronnetzer — COMMENTED** [review]
looks awesome overall 👏

**Comment by iYadgar** [comment]
It looks amazing IMO

File: dev-apps/dev-app/src/app/update-select/update-select-store.service.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+export class UpdateSelectStoreService
+  extends Ark<UpdateSelectState>
+  implements OnBeforeUpdate<UpdateSelectState>, OnAfterUpdate
+{
+  constructor() {
+    super(
+      createInitialState(),
+      withInterceptors<UpdateSelectState>({ beforeInit }),
+      withInterceptors<UpdateSelectState>({ afterInit, beforeUpdate: beforeUpdate1, afterUpdate }),
+      withInterceptors<UpdateSelectState>({ beforeUpdate: beforeUpdate2 }),
```
[the remark this answers is not in the document]
Comment (agroupp) [reply dev-apps/dev-app/src/app/update-select/update-select-store.service.ts:63]: There is no practical reason, it's just an example of multiple interceptors adding in different ways. Since it's a playground, I use it to write some ridiculous code to test most of edge cases.

File: packages/ark/src/lib/ark/entities/hooks.ts (hunk 4 above)
Comment (agroupp) [reply packages/ark/src/lib/ark/entities/hooks.ts:5]: Precisely, it is. The thing is that some developers might prefer functions, another ones angular way of lifecycle hooks. I believe it can be good idea to give customers a choice.

File: dev-apps/dev-app/src/app/update-select/update-select-store.service.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ronnetzer) [reply dev-apps/dev-app/src/app/update-select/update-select-store.service.ts:63]: and what do you say about `beforeUpdate: [beforeUpdate1, beforeUpdate2]`?

[the remark this answers is not in the document]
Comment (agroupp) [reply dev-apps/dev-app/src/app/update-select/update-select-store.service.ts:63]: I don't mind

File: .github/CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
+# Contributing to ng-console
```
Comment (ronnetzer) [inline .github/CONTRIBUTING.md:1]:
```suggestion
# Contributing to ark
```

**Review by ronnetzer — APPROVED** [review]
